tests/v2/validation/certrotation/cert_rotation.go

```diff
@@ -104,5 +150,243 @@ func rotateCerts(client *rancher.Client, clusterName string) error {
+	}
+
+	nodeList, err := client.Management.Node.List(&types.ListOpts{Filters: map[string]interface{}{
+		"clusterId": clusterID,
+	}})
+	if err != nil {
+		return err
+	}
+
+	nodeCertificates := map[string]map[string]string{}
+
+	for _, node := range nodeList.Data {
+		newCertificate, err := getCertificatesFromV3Node(client, &node)
+		if err != nil {
+			return err
+		}
+
+		nodeCertificates[node.ID] = newCertificate
+	}
+
+	cluster, err := client.Management.Cluster.ByID(clusterID)
+	if err != nil {
+		return err
+	}
+
+	_, err = client.Management.Cluster.ActionRotateCertificates(
+		cluster,
+		&management.RotateCertificateInput{CACertificates: false, Services: ""},
+	)
+	if err != nil {
+		return err
+	}
+
+	logrus.Infof("updated cluster, certs are rotating...")
+
+	adminClient, err := rancher.NewClient(client.RancherConfig.AdminToken, client.Session)
+	if err != nil {
+		return err
+	}
+
+	err = clusters.WaitClusterToBeUpgraded(adminClient, clusterID)
+	if err != nil {
+		return err
+	}
+
+	postRotatedCertificates := map[string]map[string]string{}
+
+	// kwait.PollUntilContextTimeout(context.TODO(), time.Minute, time.Minute, true, func(ctx context.Context) (done bool, err error) {
+	// 	return false, nil
+	// })
+
+	for _, node := range nodeList.Data {
+		newCertificate, err := getCertificatesFromV3Node(client, &node)
+		if err != nil {
+			return err
+		}
+
+		postRotatedCertificates[node.ID] = newCertificate
+	}
+
+	isAllCertRotated := compareCertificatesFromMachines(nodeCertificates, postRotatedCertificates)
+	if !isAllCertRotated {
+		return errors.New("certs weren't properly rotated")
+	}
+
 	return nil
 }
+
+func compareCertificatesFromMachines(certObject1, certObject2 map[string]map[string]string) bool {
+	for nodeID := range certObject1 {
+		for certType := range certObject1[nodeID] {
+
+			if certObject1[nodeID][certType] == certObject2[nodeID][certType] {
+				logrus.Infof("non-rotated cert info:")
+				logrus.Infof("%s %s was not updated: %s", nodeID, certType, certObject1[nodeID][certType])
+
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func getCertificatesFromMachine(client *rancher.Client, machineNode *v1.SteveAPIObject, sshUser string) (map[string]string, error) {
+	certificates := map[string]string{}
+
+	sshNode, err := sshkeys.GetSSHNodeFromMachine(client, sshUser, machineNode)
+	if err != nil {
+		return nil, err
+	}
+
+	logrus.Infof("Getting certificates from machine: %s", machineNode.Name)
+
+	clusterType := machineNode.Labels["node.kubernetes.io/instance-type"]
+	certsPath := "/var/lib/rancher/" + clusterType + "/server/tls/"
+
+	certsList := []string{
+		"client-admin",
+		"client-auth-proxy",
+		"client-controller",
+		"client-kube-apiserver",
+		"client-kubelet",
+		"client-kube-proxy",
+
+		"client-" + clusterType + "-cloud-controller",
+		"client-" + clusterType + "-controller",
+
+		"client-scheduler",
+		// "client-supervisor", not in either list, need to check what is expected. not currently rotated
```

Comment:
Can you remove the comments since it seems they aren't needed?

Reply:
These comments are actually waiting on confirmation from RKE2 team. I'll remove the comments shortly. tldr is, this actually found some bugs because these were supposed to be rotated but weren't.